---
title: "TCGAutils: Helper functions for working with TCGA datasets"
author: "Waldron Lab"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output:
  BiocStyle::html_document:
    number_sections: yes
    toc: true
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{TCGAutils Essentials}
  %\VignetteEncoding{UTF-8}
---

```{r, echo=FALSE, warning=FALSE}
suppressPackageStartupMessages({
    library(TCGAutils)
    library(curatedTCGAData)
    library(MultiAssayExperiment)
    library(RTCGAToolbox)
})
```

# Installation

```{r, eval=FALSE}
BiocInstaller::biocLite("TCGAutils")
```

Load some packages for the vignette:

```{r, eval=FALSE}
library(TCGAutils)
library(curatedTCGAData)
library(MultiAssayExperiment)
library(RTCGAToolbox)
```

# Overview

The `TCGAutils` package provides some utility functions for working with
data from `The Cancer Genome Atlas`. Some functions are geared towards working
with existing tools and infrastructure such as `MultiAssayExperiment` and
`curatedTCGAData`. Other functions are intended to work with flat data
structures such as `data.frame` or `DataFrame`.

# `curatedTCGAData` helpers

Functions such as `getSubtypeMap` and `getClinicalNames` allow the user to
manage data inside a downloaded `MultiAssayExperiment` object from
`curatedTCGAData`. `sampleTables` and `separateSamples` allow the user to
work with TCGA samples already present in the data.

## `MultiAssayExperiment` from `curatedTCGAData`

First, download the a part Colon Adenocarcinoma (COAD) dataset using
`curatedTCGAData` via `ExperimentHub`. The command will download any
data type that starts with `CN*` such as `CNASeq`:

```{r, echo = FALSE}
suppressMessages({
coad <- curatedTCGAData::curatedTCGAData(diseaseCode = "COAD",
    assays = "CN*", dry.run = FALSE)
})
```

```{r, eval = FALSE}
coad <- curatedTCGAData::curatedTCGAData(diseaseCode = "COAD",
    assays = "CN*", dry.run = FALSE)
```

For a list of all available data types, use `dry.run = FALSE` and an
asterisk `*` as the assay input value:

```{r}
curatedTCGAData("COAD", "*")
```

## `sampleTables`

What sample types are present in the data?

```{r}
sampleTables(coad)
```

The `sampleTables` will give you a tally of all the available
samples based on the TCGA barcode information. 

For reference, see `sampleTypes` table contained in the package.

```{r}
data("sampleTypes")
sampleTypes
```

## `separateSamples`

Now, `separateSamples` takes a vector of sample codes and partitions the current
assays into assays with an appended sample code.

```{r}
(tnmae <- separateSamples(coad, c("01", "11")))
```

To match tumors to adjacent normals, `MatchedAssayExperiment` will ensure
that every biological unit / patient is represented across all datasets.

```{r}
(matchmae <- as(tnmae, "MatchedAssayExperiment"))
```

Only about 12 participants have a matched tumor and solid normal samples.

## `getSubtypeMap` - subtype information

Subtype mapping information is saved within the `metadata` of the `colData`
slot of a `MultiAssayExperiment` downloaded from `curatedTCGAData`.
Easy access is provided by this convenience function. These mappings refer to
the curation efforts to categorize each subtype variable into a general
category.

Using the helper, one can obtain the subtype map:

```{r}
getSubtypeMap(coad)
```

## `getClinicalNames`

Many of the `curatedTCGAData` `colData` datasets contain an inordinate amount
of columns. In order to separate the standard run variables from the rest,
this function will allow the user to pull up the names from a shortened
list of variables obtained from `RTCGAToolbox` and the Firehose pipeline.
This can then be used to subset the `MultiAssayExperiment` object obtained
from `curatedTCGADAta`.

```{r}
getClinicalNames("COAD")
```

Some names may not exactly match the `colData` names in the object due
to differences in variable types. These variables are kept separate and
differentiated with `x` and `y`. For example, `vital_status` in this case
corresponds to two different variables obtained from the pipeline. One variable
is interger type and the other character:

```{r}
class(colData(coad)[["vital_status.x"]])
class(colData(coad)[["vital_status.y"]])

table(colData(coad)[["vital_status.x"]])
table(colData(coad)[["vital_status.y"]])
```

These conflicts can be resolved with further curation.

## `addClinical`

This function takes a `data.frame` or `DataFrame` and merges it with the
`colData` from an existing `MultiAssayExperiment` object. It will match
column names and row names to do a full merge of both data sets. This
convenience function is especially useful when the user has additional
subtype information to incorporate into the `colData` slot.

```{r}
addClinical(MultiAssayExperiment(), data.frame())
```

# Data to Bioconductor classes - from files or data classes

A few functions in the package accept either files or classes such as
`data.frame` and `FirehoseGISTIC` as input and return to standard Bioconductor
classes.

## `makeGRangesListFromExonFiles`

In this example we use a legacy exon quantification file from the Genomic
Data Commons. Some adjustments have to me made to the file name to account for
Windows operating system limitations. We then use
`makeGRangesListFromExonFiles` to create a `GRangesList` from vectors of file
paths and names (where necessary).

```{r}
## Load example file found in package
pkgDir <- system.file("extdata", package = "TCGAutils", mustWork = TRUE)
exonFile <- list.files(pkgDir, pattern = "cation\\.txt$", full.names = TRUE)
exonFile

## We add the original file prefix to query for the UUID and get the 
## TCGAbarcode
filePrefix <- "unc.edu.32741f9a-9fec-441f-96b4-e504e62c5362.1755371."

## Add actual file name manually
makeGRangesListFromExonFiles(exonFile,
    fileNames = paste0(filePrefix, basename(exonFile)))
```

## `makeGRangesListFromTCGA`

Alternatively, we obtain an example copy number dataset with ranged information
and use `makeGRangesListFromTCGA` to convert the data to a `GRangesList`.
The data file can be obtained from the Genomic Data Commons. This example
uses a copy number alterations dataset from Bladder Urothelial Carcinoma (BLCA). 

```{r}
grlFile <- system.file("extdata", "grlTCGA.txt", package = "TCGAutils")
grl <- read.table(grlFile)
head(grl)

makeGRangesListFromTCGA(grl, split.field = "Sample")

makeGRangesListFromTCGA(grl, split.field = "Sample", keep.extra.columns = TRUE)
```

## `makeSummarizedExperimentFromGISTIC`

The `FirehoseGISTIC` class from the `RTCGAToolbox` can be used as input
for this function. It allows the user to obtain thresholded by gene data,
probabilities and peak regions.

```{r}
tempDIR <- tempdir()
co <- getFirehoseData("COAD", clinical = FALSE, GISTIC = TRUE,
    destdir = tempDIR)

selectType(co, "GISTIC")
class(selectType(co, "GISTIC"))

makeSummarizedExperimentFromGISTIC(co, "Peaks")
```

# Identifiers

## Translation

The TCGA project has generated massive amounts of data. Some data can be
obtained with **U**niversally **U**nique **ID**entifiers (**UUID**) and other
data with TCGA barcodes. `TCGAutils` makes functions available for two-way
translation between identifiers.

### TCGA barcode to UUID

Here we take the first 2 TCGA barcodes for the copy-number alterations dataset
we downloaded previously. 

```{r}
(xbarcode <- head(colnames(coad)[["COAD_CNASeq-20160128"]], 4L))
barcodeToUUID(xbarcode)
```

### UUID to TCGA barcode

Here we have a known case UUID that we want to translate into a TCGA barcode.

```{r}
UUIDtoBarcode("ae55b2d3-62a1-419e-9f9a-5ddfac356db4", id_type = "case_id")
```

Where we have a known file UUID that we translate into the associated TCGA
barcode. Optional information can include sample, portion/analyte and
plate/center data.

```{r}
UUIDtoBarcode("0001801b-54b0-4551-8d7a-d66fb59429bf",
    id_type = "file_id", end_point = "center")
```

## Parsing

Several functions exist for working with TCGA barcodes. The main function being
`TCGAbarcode`. It takes a TCGA barcode and returns the requested data.

```{r}
## Return participant barcodes
TCGAbarcode(xbarcode, participant = TRUE)

## Just return samples
TCGAbarcode(xbarcode, participant = FALSE, sample = TRUE)

## Include sample data as well
TCGAbarcode(xbarcode, participant = TRUE, sample = TRUE)

## Include portion and analyte data
TCGAbarcode(xbarcode, participant = TRUE, sample = TRUE, portion = TRUE)
``` 

## Sample select

Sample selection is also possible with `TCGAsampleSelect`. Based on lookup
table values, the user can select certain sample types from a vector of
sample barcodes. Below we select "Primary Solid Tumors" from the supplied
barcode vector. Such function returns a logical vector corresponding to
barcodes that match the given sample code.

```{r}
## Select primary solid tumors
TCGAsampleSelect(xbarcode, "01")

## Select blood derived normals
TCGAsampleSelect(xbarcode, "10")
```

## `data.frame` representation of barcode

The straightforward `TCGAbiospec` function will take the information contained
in the TCGA barcode and display it in `data.frame` format with appropriate
column names. 

```{r}
TCGAbiospec(xbarcode)
```

# Reference data

The `TCGAutils` package contains a couple of helper datasets for defining
sample type codes that can be obtained from a TCGA barcode.

## `sampleTypes`

As shown previously, the reference dataset `sampleTypes` is available as
obtained from the TCGA website. See `?sampleTypes` for source url.

```{r}
## Obtained previously
sampleCodes <- TCGAbarcode(xbarcode, participant = FALSE, sample = TRUE)

## Lookup table
head(sampleTypes)

## Match codes found in the barcode to the lookup table
sampleTypes[match(unique(substr(sampleCodes, 1L, 2L)), sampleTypes[["Code"]]), ]
```

Source: https://gdc.cancer.gov/resources-tcga-users/tcga-code-tables/sample-type-codes

## `clinicalNames` - Firehose pipeline clinical variables

This dataset contains a list variable names accross all `colData` datasets found
in `curatedTCGAData`. This allows the user to search for particular variables
within this short list of common variable names. Shipped `curatedTCGAData`
`MultiAssayExperiment` objects contain several more variables than the ones
listed here. These variables are provided by default by the `RTCGAToolbox`
which pulls data from the Firehose pipeline.

```{r}
data("clinicalNames")

clinicalNames

lengths(clinicalNames)
```

# `sessionInfo`

```{r}
sessionInfo()
```

